netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java

```diff
@@ -112,168 +97,252 @@ public static ChannelHandler serverTls(SSLEngine sslEngine) {
+      if (!ctx.channel().isActive() || flushed) {
+        return;
       }
-    };
-    return new Negotiation() {
-      private final SettableFuture<Void> completeFuture = SettableFuture.create();
-      @Override
-      public ChannelInitializer<Channel> initializer() {
-        return initializer;
+      // Make sure that method can't be reentered, so that the ordering
+      // in the queue can't be messed up.
+      flushed = true;
+      while (!bufferedWrites.isEmpty()) {
+        ChannelWrite write = bufferedWrites.poll();
+        ctx.write(write.msg, write.promise);
       }
+      ctx.flush();
+      assert bufferedWrites.isEmpty();
+      bufferedWrites = null;
+      // Removal has to happen last as the above writes will likely trigger
+      // new writes that have to be added to the end of queue in order to not
+      // mess up the ordering.
+      ctx.pipeline().remove(this);
+    }
 
-      @Override
-      public void onConnected(Channel channel) {
-        completeFuture.set(null);
-      }
+    private static class ChannelWrite {
+      Object msg;
+      ChannelPromise promise;
 
-      @Override
-      public ListenableFuture<Void> completeFuture() {
-        return completeFuture;
+      ChannelWrite(Object msg, ChannelPromise promise) {
+        this.msg = msg;
+        this.promise = promise;
       }
-    };
+    }
   }
 
   /**
-   * Report protocol upgrade completion using a promise.
+   * Buffers all writes until the TLS Handshake is complete.
    */
-  private static class UpgradeCompletionHandler extends ChannelInboundHandlerAdapter {
-    private final SettableFuture<Void> upgradeFuture = SettableFuture.create();
+  private static class BufferUntilTlsNegotiatedHandler extends BufferWritesHandler {
 
-    public ListenableFuture<Void> getUpgradeFuture() {
-      return upgradeFuture;
+    BufferUntilTlsNegotiatedHandler(ChannelHandler... handlers) {
+      super(handlers);
     }
 
     @Override
     public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
-      if (!upgradeFuture.isDone()) {
-        if (evt == HttpClientUpgradeHandler.UpgradeEvent.UPGRADE_REJECTED) {
-          upgradeFuture.setException(new RuntimeException("HTTP/2 upgrade rejected"));
-        } else if (evt == HttpClientUpgradeHandler.UpgradeEvent.UPGRADE_SUCCESSFUL) {
-          upgradeFuture.set(null);
-          ctx.pipeline().remove(this);
+      if (evt instanceof SslHandshakeCompletionEvent) {
+        SslHandshakeCompletionEvent handshakeEvent = (SslHandshakeCompletionEvent) evt;
+        if (handshakeEvent.isSuccess()) {
+          flushBufferedAndRemove(ctx);
+        } else {
+          failBufferedAndClose(ctx);
         }
       }
+      super.userEventTriggered(ctx, evt);
+    }
+  }
+
+  /**
+   * Buffers all writes until the {@link Channel} is active.
+   */
+  private static class BufferUntilChannelActiveHandler extends BufferWritesHandler {
```

Comment:
ah ok, I didn't know the pipeline would actually let that happen ... makes sense.